android/vcmi-app/src/main/java/eu/vcmi/vcmi/util/FileUtil.java

```diff
@@ -152,4 +152,136 @@ private static String getFilenameFromUri(String sourceFileUri, Context context)
 
         return fileName;
     }
+
+	// Build a flat list of copy tasks for a SAF tree selection.
+	// Output rows are "srcUri\tTarget\tName" where Target is one of Data/Maps/Mp3 decided by file extension.
+	// Heuristic: if user picked "Data" and we found no Maps/Mp3, also scan ../Maps and ../Mp3 one level up.
+	public static String[] findFilesForCopy(String treeUriStr, android.content.Context ctx) {
+	    // Build a DocumentFile tree from the SAF tree URI
+	    final android.net.Uri treeUri = android.net.Uri.parse(treeUriStr);
+	    final androidx.documentfile.provider.DocumentFile root =  androidx.documentfile.provider.DocumentFile.fromTreeUri(ctx, treeUri);
+	    if (root == null) return new String[0];
+	
+	    final java.util.ArrayList<String> out = new java.util.ArrayList<>();
+	    int foundMaps = 0, foundMp3 = 0;
+	
+	    // 1) Traverse the selected tree, collect files by extension
+	    {
+	        final java.util.ArrayDeque<androidx.documentfile.provider.DocumentFile> stack = new java.util.ArrayDeque<>();
```

Comment:
if you need a stack, use `java.util.Stack`, otherwise call the variable appropriately